### PR TITLE
Port allmydata.test.test_storage to Python 3

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3397

### CODE REVIEW

File: src/allmydata/storage/server.py [hunk 1]
```diff
@@ -676,6 +677,10 @@ def remote_slot_readv(self, storage_index, shares, readv):
 
     def remote_advise_corrupt_share(self, share_type, storage_index, shnum,
                                     reason):
+        # This is a remote API, I believe, so this has to be bytes for legacy
+        # protocol backwards compatibility reasons.
```
Comment (exarkun) [inline src/allmydata/storage/server.py:681]: :+1:

File: src/allmydata/test/test_storage.py [hunk 2]
```diff
@@ -1,3 +1,13 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:5]: This should have a docstring that at least states the source file has been ported to Python 3.

File: src/allmydata/test/test_storage.py [hunk 3]
```diff
@@ -65,32 +75,32 @@ def make_lease(self):
         cancel_secret = os.urandom(32)
         expiration_time = time.time() + 5000
         return LeaseInfo(owner_num, renew_secret, cancel_secret,
-                         expiration_time, "\x00" * 20)
+                         expiration_time, b"\x00" * 20)
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:78]: Perhaps worth adding docs or a type check to `LeaseInfo` for this parameter (`nodeid`?) to make sure it is consistently bytes?

Assuming all callers are actually supplying bytes currently...

File: src/allmydata/test/test_storage.py [hunk 4]
```diff
@@ -302,7 +312,7 @@ def workdir(self, name):
 
     def create(self, name, reserved_space=0, klass=StorageServer):
         workdir = self.workdir(name)
-        ss = klass(workdir, "\x00" * 20, reserved_space=reserved_space,
+        ss = klass(workdir, b"\x00" * 20, reserved_space=reserved_space,
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:315]: Maybe docs/typecheck for `StorageServer` `nodeid` parameter too.

File: src/allmydata/test/test_storage.py [hunk 5]
```diff
@@ -1302,56 +1318,58 @@ def secrets(n):
...
-        self.failUnlessEqual(readv("si1", [], [(0,10)]),
+        self.failUnlessEqual(readv(b"si1", [], [(0,10)]),
                              {})
         # and the bucket directory should now be gone
-        si = base32.b2a("si1")
+        si = base32.b2a(b"si1")
         # note: this is a detail of the storage server implementation, and
         # may change in the future
+        if PY3:
+            si = si.decode("utf-8")
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:1372]: Does a use of `bytes_to_native_string` make sense here?

File: src/allmydata/test/test_storage.py [hunk 6]
```diff
@@ -1392,7 +1410,7 @@ def test_get_slot_leases_empty_slot(self):
         When ``get_slot_leases`` is called for a slot for which the server has no
         shares, it returns an empty iterable.
         """
-        ss = self.create(b"test_get_slot_leases_empty_slot")
+        ss = self.create("test_get_slot_leases_empty_slot")
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:1413]: What a delightful break in the monotony. :)

File: src/allmydata/test/test_storage.py [hunk 7]
```diff
@@ -1471,15 +1489,19 @@ def tearDown(self):
...
+        if isinstance(tag, int):
+            tag = b"%d" % tag
+        return hashutil.tagged_hash(b"renew_blah", tag)
 
 
     def cancel_secret(self, tag):
-        return hashutil.tagged_hash("cancel_blah", str(tag))
+        if isinstance(tag, int):
+            tag = b"%d" % tag
+        return hashutil.tagged_hash(b"cancel_blah", tag)
```
Comment (exarkun) [inline src/allmydata/test/test_storage.py:1504]: More implementations of these helpers?  Oh.   Yea.  More.  Okay...

**Review by exarkun — APPROVED** [review]
Thanks.  Some inline comments.  One more thing (that I can't do inline) - I think neither the tests nor implementation of "remote_get_version" have been ported.  I suppose they pass because Foolscap happily passes the strings as bytes or unicode and the tests then happily access the result using the same type just by running on the same version of Python as the server code.

If you want to port that in a follow-up rather than in this PR that's fine with me.  It'll probably be easier to review that way, too.

File: src/allmydata/test/test_storage.py (hunk 3 above)
Comment (itamarst) [reply src/allmydata/test/test_storage.py:78]: Looks like it already has assert for that.

File: src/allmydata/test/test_storage.py (hunk 5 above)
Comment (itamarst) [reply src/allmydata/test/test_storage.py:1372]: Good idea.

**Comment by itamarst** [comment]
Created https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3402 for remote_get_version.

**Pull request merged**